src/clang-cpp-frontend/clang_cpp_convert.cpp

```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
 
   return false;
 }
+
+bool clang_cpp_convertert::annotate_class_field_access(
+  const clang::FieldDecl *field,
+  struct_typet::componentt &comp)
+{
+  std::string access;
+  if(get_access_from_decl(field, access))
+    return true;
+
+  // annotate access in component
+  comp.set_access(access);
+  return false;
+}
+
+bool clang_cpp_convertert::get_access_from_decl(
+  const clang::Decl *decl,
+  std::string &access)
+{
+  switch(decl->getAccess())
+  {
+  case clang::AS_public:
+  {
+    access = "public";
+    break;
+  }
+  case clang::AS_private:
+  {
+    access = "private";
+    break;
+  }
+  case clang::AS_protected:
+  {
+    access = "protected";
+    break;
+  }
+  default:
+  {
+    log_error("Unknown specifier returned from clang in {}", __func__);
+    return true;
+  }
+  }
+
+  return false;
+}
+
+bool clang_cpp_convertert::annotate_cpp_methods(
+  const clang::CXXMethodDecl *cxxmdd,
+  exprt &new_expr)
+{
+  code_typet &component_type = to_code_type(new_expr.type());
+
+  // annotate ctor and dtor return type
+  if(cxxmdd->getKind() == clang::Decl::CXXDestructor)
+  {
+    typet rtn_type("destructor");
+    component_type.return_type() = rtn_type;
+  }
+  if(cxxmdd->getKind() == clang::Decl::CXXConstructor)
+  {
+    typet rtn_type("constructor");
+    component_type.return_type() = rtn_type;
+  }
+
+  // annotate parent
+  std::string parent_class_name = getFullyQualifiedName(
+    ASTContext->getTagDeclType(cxxmdd->getParent()), *ASTContext);
+  std::string parent_class_id = tag_prefix + parent_class_name;
+  component_type.set("#member_name", parent_class_id);
+
+  // annotate name
+  std::string method_id, method_name;
+  clang_c_convertert::get_decl_name(*cxxmdd, method_name, method_id);
+  new_expr.name(method_id);
+
+  // annotate access
+  std::string access;
+  if(get_access_from_decl(cxxmdd, access))
+    return true;
+  new_expr.set("access", access);
+
+  // annotate base name
+  new_expr.base_name(parent_class_name);
+
+  // annotate pretty name
+  new_expr.pretty_name(parent_class_name);
+
+  // annotate inline
+  new_expr.set("is_inlined", cxxmdd->isInlined());
+
+  // annotate location
+  locationt location_begin;
+  get_location_from_decl(*cxxmdd, location_begin);
+  new_expr.location() = location_begin;
+
+  // We still need to add a trivial ctor/dtor as a `component` in class symbol's type
+  // remove "statement: skip" otherwise it won't be added
+  // TODO-split: Any more elegant way to do it?
```

Comment:
Can you explain why we need this code?

Reply:
Preparation for vptr and vtable extensions. We need to make some implicit code for vptr later - sorry it may not be obvious for the time being.